File: quickstep/src/com/android/launcher3/taskbar/NudgeView.java
/*
 * Copyright (C) 2025 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.launcher3.taskbar;

import static com.android.launcher3.anim.AnimatedFloat.VALUE;
import static com.android.launcher3.anim.AnimatorListeners.forEndCallback;
import static com.android.launcher3.taskbar.StashedHandleViewController.ALPHA_INDEX_NUDGED;
import static com.android.launcher3.taskbar.StashedHandleViewController.ALPHA_INDEX_STASHED;

import android.animation.AnimatorSet;
import android.animation.ValueAnimator;
import android.content.Context;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;
import android.view.ViewGroup;
import android.widget.ImageView;

import androidx.annotation.Nullable;

import com.android.launcher3.Flags;
import com.android.launcher3.R;
import com.android.launcher3.anim.AnimatedFloat;
import com.android.launcher3.anim.SpringAnimationBuilder;
import com.android.launcher3.util.MultiValueAlpha;

public class NudgeView extends ImageView {

    private static final float DAMPING_RATIO = 0.6f;
    private static final float STIFFNESS = 380f;

    private AnimatorSet mAnimatorSet;
    private boolean mNudgeShown;
    private int mNudgePillWidth;
    private int mNudgePillHeight;
    private int mStashedHandleWidth;
    private int mStashedHandleHeight;
    private ImageView mNudgeIcon;
    private MultiValueAlpha mStashedHandleViewAlpha;

    private final AnimatedFloat mAlphaForHandleView = new AnimatedFloat(
            this::updateAlphaForHandleView);
    private final AnimatedFloat mAlphaForNudgeIcon = new AnimatedFloat(
            this::updateAlphaForNudgeIcon);
    private final AnimatedFloat mWidthForNudgePill = new AnimatedFloat(
            this::updateWidthForNudge);
    private final AnimatedFloat mHeightForNudgePill = new AnimatedFloat(
            this::updateHeightForNudge);

    public NudgeView(Context context) {
        this(context, null);
    }

    public NudgeView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public NudgeView(Context context, AttributeSet attrs, int defStyleAttr) {
        this(context, attrs, defStyleAttr, 0);
    }

    public NudgeView(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        if (Flags.nudgePill()) {
            mNudgePillWidth = context.getResources().getDimensionPixelSize(
                    R.dimen.taskbar_nudge_pill_width);
            mNudgePillHeight = context.getResources().getDimensionPixelSize(
                    R.dimen.taskbar_nudge_pill_height);
            mStashedHandleWidth = context.getResources().getDimensionPixelSize(
                    R.dimen.taskbar_nudge_pill_width);
            mStashedHandleHeight = context.getResources().getDimensionPixelSize(
                    R.dimen.taskbar_stashed_handle_height);
        }
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();
        mNudgeIcon = findViewById(R.id.nudge_icon);
    }

    void updateNudgeIcon(boolean showNudge, @Nullable Drawable icon,
            @Nullable MultiValueAlpha stashedHandleViewAlpha) {
        if (!Flags.nudgePill() || stashedHandleViewAlpha == null) {
            return;
        }
        mStashedHandleViewAlpha = stashedHandleViewAlpha;
        // Set update visibility to false when showing the nudge otherwise
        // StashHandleView will be INVISIBLE at low alpha resulting in being unable to invoke CtS.
        mStashedHandleViewAlpha.setUpdateVisibility(false);
        mNudgeIcon.setImageDrawable(null);
        if (icon != null) {
            mNudgeIcon.setImageDrawable(icon);
        }
        mNudgeIcon.setVisibility(VISIBLE);
        if (showNudge && !mNudgeShown) {
            post(() -> animationBarToPill(true /* showNudge */));
        } else if (!showNudge){
            post(() -> animationBarToPill(false /* showNudge */));
        }
    }

    private void animationBarToPill(boolean showNudge) {
        if (!Flags.nudgePill()) {
            return;
        }
        if (mAnimatorSet != null && mAnimatorSet.isRunning()) {
            mAnimatorSet.cancel();
        }
        mNudgeShown = showNudge;
        // Alpha
        float targetIconAlpha = showNudge ? 1.0f : 0;
        float targetHandleViewAlpha = showNudge ? 0 : 1.0f;
        // Animate from navBar -> nudgePill if showNudge is true.
        float startWidth = showNudge ? mStashedHandleWidth : mNudgePillWidth;
        float startHeight = showNudge ? mStashedHandleHeight : mNudgePillHeight;
        float endWidth = showNudge ? mNudgePillWidth : mStashedHandleWidth;
        float endHeight = showNudge ? mNudgePillHeight : mStashedHandleHeight;

        ValueAnimator alphaNudgeIcon = new SpringAnimationBuilder(mContext)
                .setStartValue(mNudgeIcon.getAlpha())
                .setEndValue(targetIconAlpha)
                .setDampingRatio(DAMPING_RATIO)
                .setStiffness(STIFFNESS)
                .build(mAlphaForNudgeIcon, VALUE);
        if (!showNudge) {
            alphaNudgeIcon.addListener(forEndCallback(() -> {
                mNudgeIcon.setVisibility(GONE);
                mStashedHandleViewAlpha.setUpdateVisibility(true);
            }));
        }
        ValueAnimator alphaHandleView = new SpringAnimationBuilder(mContext)
                .setStartValue(mStashedHandleViewAlpha.get(ALPHA_INDEX_STASHED).getValue())
                .setEndValue(targetHandleViewAlpha)
                .setDampingRatio(DAMPING_RATIO)
                .setStiffness(STIFFNESS)
                .build(mAlphaForHandleView, VALUE);
        ValueAnimator widthAnimation = new SpringAnimationBuilder(mContext)
                .setStartValue(startWidth)
                .setEndValue(endWidth)
                .setDampingRatio(DAMPING_RATIO)
                .setStiffness(STIFFNESS)
                .build(mWidthForNudgePill, VALUE);
        ValueAnimator heightAnimation = new SpringAnimationBuilder(mContext)
                .setStartValue(startHeight)
                .setEndValue(endHeight)
                .setDampingRatio(DAMPING_RATIO)
                .setStiffness(STIFFNESS)
                .build(mHeightForNudgePill, VALUE);
        mAnimatorSet = new AnimatorSet();
        mAnimatorSet.playTogether(
                alphaNudgeIcon, alphaHandleView,
                widthAnimation, heightAnimation
        );
        mAnimatorSet.start();
    }

    private void updateAlphaForHandleView() {
        if (!Flags.nudgePill()) {
            return;
        }
        float alpha = mAlphaForHandleView.value;
        mStashedHandleViewAlpha.get(ALPHA_INDEX_NUDGED).setValue(alpha);
    }

    private void updateAlphaForNudgeIcon() {
        if (!Flags.nudgePill()) {
            return;
        }
        float alpha = mAlphaForNudgeIcon.value;
        mNudgeIcon.setAlpha(alpha);
    }

    private void updateWidthForNudge() {
        if (!Flags.nudgePill()) {
            return;
        }
        float width = mWidthForNudgePill.value;
        ViewGroup.MarginLayoutParams nudgePillLayoutParams =
                (ViewGroup.MarginLayoutParams) mNudgeIcon.getLayoutParams();
        nudgePillLayoutParams.width = (int) width;
        mNudgeIcon.setLayoutParams(nudgePillLayoutParams);
    }

    private void updateHeightForNudge() {
        if (!Flags.nudgePill()) {
            return;
        }
        float height = mHeightForNudgePill.value;
        ViewGroup.MarginLayoutParams nudgePillLayoutParams =
                (ViewGroup.MarginLayoutParams) mNudgeIcon.getLayoutParams();
        nudgePillLayoutParams.height = (int) height;
        mNudgeIcon.setLayoutParams(nudgePillLayoutParams);
    }
}

File: quickstep/src/com/android/launcher3/taskbar/growth/NudgeView.kt
/*
 * Copyright (C) 2025 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.launcher3.taskbar.growth

import android.animation.AnimatorSet
import android.animation.ValueAnimator
import android.content.Context
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.ViewGroup
import android.view.animation.Interpolator
import android.window.OnBackInvokedDispatcher
import androidx.core.view.updateLayoutParams
import com.android.app.animation.Interpolators
import com.android.launcher3.AbstractFloatingView
import com.android.launcher3.R
import com.android.launcher3.anim.AnimatorListeners
import com.android.launcher3.views.ActivityContext

private const val ENTER_DURATION_MS = 300L
private const val EXIT_DURATION_MS = 150L

/** Floating nudge. */
class NudgeView
@JvmOverloads
constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) :
    AbstractFloatingView(context, attrs, defStyleAttr) {

    private val activityContext: ActivityContext = ActivityContext.lookupContext(context)

    private val enterYDelta = resources.getDimension(R.dimen.nudge_enter_y_delta)
    private val exitYDelta = resources.getDimension(R.dimen.nudge_exit_y_delta)

    /** Container where the nudge's body should be inflated. */
    lateinit var content: ViewGroup
        private set

    /** Callback invoked when the nudge is being closed. */
    var onCloseCallback: () -> Unit = {}
    private var openCloseAnimator: AnimatorSet? = null
    /** Used to set whether users can tap outside the current nudge window to dismiss it */
    var allowTouchDismissal = true

    /** Animates the nudge into view. */
    fun show() {
        if (isOpen) {
            return
        }

        mIsOpen = true
        activityContext.dragLayer.addView(this)

        // Make sure we have enough height to display all of the content, which can be an issue on
        // large text and display scaling configurations. If we run out of height, remove the width
        // constraint to reduce the number of lines of text and hopefully free up some height.
        activityContext.dragLayer.measure(MeasureSpec.UNSPECIFIED, MeasureSpec.UNSPECIFIED)
        if (
            measuredHeight + activityContext.deviceProfile.taskbarProfile.height >=
                activityContext.deviceProfile.deviceProperties.availableHeightPx
        ) {
            updateLayoutParams { width = LayoutParams.MATCH_PARENT }
        }

        openCloseAnimator = createOpenCloseAnimator(isOpening = true).apply { start() }
    }

    override fun onFinishInflate() {
        super.onFinishInflate()

        content = requireViewById(R.id.content)
    }

    override fun handleClose(animate: Boolean) {
        if (!isOpen) {
            return
        }

        onCloseCallback()
        if (!animate) {
            return closeComplete()
        }

        openCloseAnimator?.cancel()
        openCloseAnimator = createOpenCloseAnimator(isOpening = false)
        openCloseAnimator?.apply {
            addListener(AnimatorListeners.forEndCallback(this@NudgeView::closeComplete))
            start()
        }
    }

    override fun isOfType(type: Int): Boolean = type and TYPE_NUDGE != 0

    override fun onControllerInterceptTouchEvent(ev: MotionEvent?): Boolean {
        if (
            ev?.action == MotionEvent.ACTION_DOWN &&
                !activityContext.dragLayer.isEventOverView(this, ev) &&
                allowTouchDismissal
        ) {
            close(true)
        }
        return false
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        findOnBackInvokedDispatcher()
            ?.registerOnBackInvokedCallback(OnBackInvokedDispatcher.PRIORITY_DEFAULT, this)
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        findOnBackInvokedDispatcher()?.unregisterOnBackInvokedCallback(this)
        // TODO: b/396507770 - investigate and coordinate with letterbox flow.
    }

    private fun closeComplete() {
        openCloseAnimator?.cancel()
        openCloseAnimator = null
        mIsOpen = false
        activityContext.dragLayer.removeView(this)
    }

    private fun createOpenCloseAnimator(isOpening: Boolean): AnimatorSet {
        val duration: Long
        val alphaValues: FloatArray
        val translateYValues: FloatArray
        val fadeInterpolator: Interpolator
        val translateYInterpolator: Interpolator

        if (isOpening) {
            duration = ENTER_DURATION_MS
            alphaValues = floatArrayOf(0f, 1f)
            translateYValues = floatArrayOf(enterYDelta, 0f)
            fadeInterpolator = Interpolators.STANDARD
            translateYInterpolator = Interpolators.EMPHASIZED_DECELERATE
        } else {
            duration = EXIT_DURATION_MS
            alphaValues = floatArrayOf(1f, 0f)
            translateYValues = floatArrayOf(0f, exitYDelta)
            fadeInterpolator = Interpolators.EMPHASIZED_ACCELERATE
            translateYInterpolator = Interpolators.EMPHASIZED_ACCELERATE
        }

        val fade =
            ValueAnimator.ofFloat(*alphaValues).apply {
                interpolator = fadeInterpolator
                addUpdateListener {
                    val alpha = it.animatedValue as Float
                    content.alpha = alpha
                }
            }

        val translateY =
            ValueAnimator.ofFloat(*translateYValues).apply {
                interpolator = translateYInterpolator
                addUpdateListener {
                    val translationY = it.animatedValue as Float
                    content.translationY = translationY
                }
            }

        return AnimatorSet().apply {
            this.duration = duration
            playTogether(fade, translateY)
        }
    }
}
